### ISSUE TITLE
PC wakes up keyboard not working

### ISSUE THREAD

**Issue by CyprienRicque** [kmonad/kmonad#753]
Hello,
When the PC wakes up from sleep, sometimes the keys are correctly received by kmonad (displayed with -ldebug) but they are not transmitted back to the PC. So I can't use the keyboard anymore.

## Config
macOS Sonoma 14.0
M1 Pro

**Comment by nightscape** [comment]
Same here. I noticed that you can make kmonad crash by pressing the `^` key left of `1`.
I've built a launcher script around kmonad that restarts it after a crash:
```bash
#!/bin/bash

# Path to the kmonad binary
KMONAD_PATH="/Users/martin/.local/bin/kmonad"
KMONAD_CONFIG_PATH="/Users/martin/Workspaces/devtools/kmonad-contrib/keymaps/nightscape/bone_mac.kbd"

# Counter for how many times kmonad has crashed
CRASH_COUNT=0

# Timestamp of the first crash in the current sequence
FIRST_CRASH_TIMESTAMP=0

while true; do
    # Run kmonad with superuser permissions
	echo "Starting kmonad"
	echo "sudo $KMONAD_PATH $KMONAD_CONFIG_PATH"
    sudo $KMONAD_PATH $KMONAD_CONFIG_PATH

    # If kmonad exits, it has crashed. Increment the crash counter.
    CRASH_COUNT=$((CRASH_COUNT + 1))

    # If this is the first crash in the current sequence, record the timestamp
    if [ $CRASH_COUNT -eq 1 ]; then
        FIRST_CRASH_TIMESTAMP=$(date +%s)
    fi

    # Get the current timestamp
    CURRENT_TIMESTAMP=$(date +%s)

    # Calculate the time difference between the first crash and now
    TIME_DIFF=$((CURRENT_TIMESTAMP - FIRST_CRASH_TIMESTAMP))

    # If kmonad has crashed 3 times in less than 60 seconds, exit the loop
    if [ $CRASH_COUNT -ge 3 ] && [ $TIME_DIFF -lt 60 ]; then
        echo "kmonad has crashed too frequently. Exiting."
        exit 1
    fi

    # If more than 60 seconds have passed since the first crash, reset the crash counter
    if [ $TIME_DIFF -ge 60 ]; then
        CRASH_COUNT=0
    fi

    # Wait for a second before trying to restart kmonad
    sleep 1
done
```

So every time I resume my MacBook, I press `^` and then wait for kmonad to respawn. Not optimal but it alleviates the pain somewhat.

**Comment by CyprienRicque** [comment]
Interesting, on my side I don't manage to make it crash using `^` 

For the time mine crashed but I am not sure what caused it.

```
Received event: Press <f2>
Running hooks
Registering untimed hook: 1063
Emitting: Press <bru>
Encountered error in KeySource: [Cannot interpret mac switch: 4,Cannot translate from mac keycode: (255,4294967295)]
Closing process: emitter_proc
Closing KeySource
Closing IOKit devices
Closing KeySink
Closing Mac key sink
kmonad: [Cannot interpret mac switch: 4,Cannot translate from mac keycode: (255,4294967295)]
```

**Comment by dalewking** [comment]
I have seen this on MacOS as well

**Comment by dalewking** [comment]
I would definitely like to see this fixed and am a developer, but am totally clueless with Haskell, though I do know C/C++. I have a way to reproduce it very reliably. My keyboard is connected via USB to a Thunderbolt display connected to my MacBook with a Thunderbolt to USB adapter. When the computer wakes up from sleep it end up in this state every time.

I tried running with -ldebug and here is what I see:

When the computer wakes up I see:
```
Tried cancelling expired hook: 1843
Tried cancelling expired hook: 1851
Tried cancelling expired hook: 1847
Tried cancelling expired hook: 1860
Tried cancelling expired hook: 1855
Tried cancelling expired hook: 1896
Tried cancelling expired hook: 1910
Tried cancelling expired hook: 1903
Tried cancelling expired hook: 1884
Tried cancelling expired hook: 1874
Tried cancelling expired hook: 1879
Tried cancelling expired hook: 1869
```
Then I press a single key and the output is absolutely flooded with thousands of messages like this:
```
--------------------------------------------------------------------------------
Received event: Press <a>
Running hooks
Block level set to: 1
Registering untimed hook: 1993
Registering 300ms hook: 1992

--------------------------------------------------------------------------------
Received event: Release <a>
Running hooks
Emitting: Press <a>
Unblocking input stream, no stored events
Emitting: Release <a>

--------------------------------------------------------------------------------
Received event: Press <a>
Running hooks
Block level set to: 1
Registering untimed hook: 1995
Registering 300ms hook: 1994

--------------------------------------------------------------------------------
Received event: Release <a>
Running hooks
Emitting: Press <a>
Unblocking input stream, no stored events
Emitting: Release <a>
```

There are occasional batches of `Tried cancelling Expired Hook`

Is that of any help?

**Comment by brttbndr** [comment]
I am on macOS Sonoma and seeing this issue; on resume from sleep, tapping a key makes KMonad peg CPU and lots of log output (at debug log-level) but UI is unresponsive / not receiving keystrokes.

I have an informed guess about the cause.

I am using the following to debug:

```
sudo kmonad ~/.config/kmonad.kbd --log-level debug
```

and in my kmonad.kbd

```
(defcfg
  ;; input  (iokit-name "Apple Internal Keyboard / Trackpad")
  input  (iokit-name)
  output (kext)
  fallthrough true
  allow-cmd true
)
```

I observe 2 things.

1. The output seems repetitive -- we're seeing the exact same Received / Emitting / Emitting
2. If I specify the optional name param to iokit-name -- then I can't reproduce. (I'm not entirely confident this is true ... but it's the case for the several attempts I've made.)

So my guess is that after resuming from sleep, KMonad is receiving the events it emits (in a way that doesn't happen before sleep?). They are reprocessed (infinitely?) and not handed off to the Karabiner kext (or dext, in my case).

It might help confirm or invalidate this theory if anyone experiencing this issue could check whether they have `input (iokit-name)` or `input (iokit-name "Some Keyboard")`. And if the former -- does supplying your keyboard name workaround / mitigate?

**Comment by brttbndr** [comment]
It looks like any "Karabiner VirtualHIDKeyboard" is excluded by name (`CFStringCompare` returns `kCFCompareEqualTo`). 

https://github.com/kmonad/kmonad/blob/1c4641f063130165516789e842629e437082c5b8/c_src/mac/keyio_mac.hpp#L78-L97

But on my machine the keyboards are (while kmonad is running, to get the virtual kbd)

```
kmonad/c_src/mac
❯ ./list-keyboards
Apple Internal Keyboard / Trackpad
Karabiner DriverKit VirtualHIDKeyboard 1.7.0
```

So I suppose the device name changed w/the migration to DriverKit.

This patch works for me, and should be backwards compatible -- match on the prefix "Karabiner " rather than the exact string "Karabiner VirtualHIDKeyboard"

```
diff --git a/c_src/mac/keyio_mac.hpp b/c_src/mac/keyio_mac.hpp
index 3088345..38de94c 100644
--- a/c_src/mac/keyio_mac.hpp
+++ b/c_src/mac/keyio_mac.hpp
@@ -75,7 +75,8 @@ void open_matching_devices(char *product, io_iterator_t iter) {
             return;
         }
     }
-    CFStringRef cfkarabiner = CFStringCreateWithCString(kCFAllocatorDefault, "Karabiner VirtualHIDKeyboard", CFStringGetSystemEncoding());
+    // assume any "Karabiner ..." is a virtual device and should be ignored
+    CFStringRef cfkarabiner = CFStringCreateWithCString(kCFAllocatorDefault, "Karabiner ", CFStringGetSystemEncoding());
     if(cfkarabiner == NULL) {
         print_iokit_error("CFStringCreateWithCString");
         if(product) {
@@ -89,7 +90,7 @@ void open_matching_devices(char *product, io_iterator_t iter) {
             print_iokit_error("IORegistryEntryCreateCFProperty");
             continue;
         }
-        bool match = (CFStringCompare(cfcurr, cfkarabiner, 0) != kCFCompareEqualTo);
+        bool match = !CFStringHasPrefix(cfcurr, cfkarabiner);
         if(product) {
             match = match && (CFStringCompare(cfcurr, cfproduct, 0) == kCFCompareEqualTo);
         }
```

**Comment by romankonz** [comment]
> 2. If I specify the optional name param to iokit-name -- then I can't reproduce. (I'm not entirely confident this is true ... but it's the case for the several attempts I've made.)

I can confirm the behaviour, setting the iokit-name fixed it for me 👍

**Comment by brttbndr** [comment]
@dalewking I submitted a PR on 27 Dec -- https://github.com/kmonad/kmonad/pull/781

For the more-than-1-keyboard use case, I'm using an approach inspired by the FAQ

https://github.com/kmonad/kmonad/blob/master/doc/faq.md#q-how-do-i-use-the-same-layout-definition-for-different-keyboards

But I actually want a _different_ layout depending on keyboard, e.g. to swap FN and LCTL (only on the Apple keyboard). So I'm using this script, spawned from launchd.

(I don't see a way to apply variant maps to different keyboards simultaneously, which I would do in Karabiner with device conditions. So I've denormalized to multiple slightly different config files, and only use strictly one keyboard at a time...)

```
#!/usr/bin/env bash
# select KMonad config based on most-preferred of attached keyboards,
# then run KMonad with that config
#
# re-executes when KMonad exits (crashes),
# handling sleep/resume and hardware changes
#

CONFDIR="/Users/brttbndr/.config/kmonad"
BINDIR="/Users/brttbndr/.local/bin"

# list-keyboards is a util that comes with kmonad,
# output modified to replace spaces with underscores
KEEBS="$($BINDIR/list-keyboards 2>&1 | tr ' ' '_')"

# the expected default case is the Apple internal keyboard,
# and if it's not present, 🤷
case "$KEEBS" in
*Adv360_Pro*)
	KB=Adv360_Pro
	;;
*Freestyle_Edge*)
	KB=Freestyle_Edge
	;;
*Apple_Internal*
	KB=Apple_Internal
	;;
*)
	echo "Unknown - $KEEBS"
	exit 1
	;;
esac

echo "Using config $KB"

"$BINDIR/kmonad" "$CONFDIR/$KB.kbd"

echo "KMonad exited, re-execing"

exec "$@"
```

**Comment by slotThe** [comment]
Closed by #781
